opslevel/resource_opslevel_service.go

```diff
@@ -1,360 +1,394 @@
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	serviceUpdateInput := opslevel.ServiceUpdateInput{
+		Description:    data.Description.ValueStringPointer(),
+		Framework:      data.Framework.ValueStringPointer(),
+		Id:             opslevel.NewID(data.Id.ValueString()),
+		Language:       data.Language.ValueStringPointer(),
+		LifecycleAlias: data.LifecycleAlias.ValueStringPointer(),
+		Name:           data.Name.ValueStringPointer(),
+		Product:        data.Product.ValueStringPointer(),
+		TierAlias:      data.TierAlias.ValueStringPointer(),
+	}
+	service, err := r.client.UpdateService(serviceUpdateInput)
+	if err != nil {
+		resp.Diagnostics.AddError("opslevel client error", fmt.Sprintf("Unable to update service, got error: %s", err))
+		return
+	}
+
+	givenAliases, diags := ListValueToStringSlice(ctx, data.Aliases)
+	if diags != nil && diags.HasError() {
+		resp.Diagnostics.AddError("Config error", fmt.Sprintf("Unable to handle given service aliases: '%s'", data.Aliases))
+		return
+	}
+	err = reconcileServiceAliases(*r.client, givenAliases, service)
+	if err != nil {
+		resp.Diagnostics.AddError("opslevel client error", fmt.Sprintf("Unable to create service aliases: '%s'", givenAliases))
+		return
+	}
+
+	givenTags, diags := ListValueToStringSlice(ctx, data.Tags)
+	if diags != nil && diags.HasError() {
+		resp.Diagnostics.AddError("Config error", fmt.Sprintf("Unable to handle given service tags: '%s'", data.Tags))
+		return
+	}
+	err = reconcileTags(*r.client, givenTags, service)
+	if err != nil {
+		resp.Diagnostics.AddError("opslevel client error", fmt.Sprintf("Unable to create service tags: '%s'", givenTags))
+		return
+	}
+
+	updatedServiceResourceModel, diags := NewServiceResourceModel(ctx, *service)
+	updatedServiceResourceModel.LastUpdated = timeLastUpdated()
+	resp.Diagnostics.Append(diags...)
+
+	tflog.Trace(ctx, "updated a service resource")
+	resp.Diagnostics.Append(resp.State.Set(ctx, &updatedServiceResourceModel)...)
+}
+
+func (r *ServiceResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var data ServiceResourceModel
+
+	// Read Terraform prior state data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	err := r.client.DeleteService(data.Id.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to delete service, got error: %s", err))
+		return
+	}
+	tflog.Trace(ctx, "deleted a service resource")
+}
+
+func (r *ServiceResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
+}
+
+// Assigns new aliases from terraform config to service and deletes aliases not in config
+func reconcileServiceAliases(client opslevel.Client, aliasesFromConfig []string, service *opslevel.Service) error {
```

Comment:
Can this just be a member of `ServiceResource`? That way you can do `r.client` instead of needing to pass it in as an arg

Reply:
The only thing needed in this function from `ServiceResource` is the client. All of `ServiceResource` is heavy handed IMO and can even be distracting.
Also while this is specific to a `service` as is, I want to keep this open to generalizing to the other resources. Possibly something similar to what we do with [tags in opslevel-go](https://github.com/OpsLevel/opslevel-go/blob/main/tags.go#L39)